# 1. Import Library

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

# 2. Load Data

In [2]:
df = pd.read_csv('Data/part4_tmdb_results_all.csv.gz', lineterminator='\n')
df.head(2)

adult                     backdrop_path belongs_to_collection      budget  \
0    NaN                               NaN                   NaN         NaN   
1    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg                   NaN  10000000.0   

                                              genres homepage       id  \
0                                                NaN      NaN      NaN   
1  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN  62127.0   

  original_language   original_title  \
0               NaN              NaN   
1                en  The Fantasticks   

                                            overview  ...  revenue runtime  \
0                                                NaN  ...      NaN     NaN   
1  Two rural teens sing and dance their way throu...  ...      0.0    86.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                                         tagline            title  video  \
0                                            NaN              NaN    NaN   
1  Try to remember the first time magic happened  The Fantasticks    0.0   

  vote_average vote_count certification  
0          NaN        NaN           NaN  
1          5.5       22.0           NaN  

[2 rows x 25 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63148 entries, 0 to 63147
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  63126 non-null  float64
 1   backdrop_path          40883 non-null  object 
 2   belongs_to_collection  4257 non-null   object 
 3   budget                 63126 non-null  float64
 4   genres                 63126 non-null  object 
 5   homepage               15142 non-null  object 
 6   id                     63126 non-null  float64
 7   original_language      63126 non-null  object 
 8   original_title         63126 non-null  object 
 9   overview               61824 non-null  object 
 10  popularity             63126 non-null  float64
 11  poster_path            58195 non-null  object 
 12  production_companies   63126 non-null  object 
 13  production_countries   63126 non-null  object 
 14  release_date           62044 non-null  object 
 15  re

# 3. Clean Data

## 3.1 Check & remove duplicated rows

In [4]:
#check duplicates
df.duplicated().sum()

21

In [5]:
#remove duplicates
df = df.drop_duplicates()

In [6]:
#check that the duplicates have been dropped
df.duplicated().sum()

0

## 3.2 Check for missing values

In [7]:
#check missing values in each column
df.isna().sum()

adult                        1
backdrop_path            22244
belongs_to_collection    58870
budget                       1
genres                       1
homepage                 47985
id                           1
original_language            1
original_title               1
overview                  1303
popularity                   1
poster_path               4932
production_companies         1
production_countries         1
release_date              1083
revenue                      1
runtime                      1
spoken_languages             1
status                       1
tagline                  38981
title                        1
video                        1
vote_average                 1
vote_count                   1
certification            47885
dtype: int64

**Observation**: There are many missing values in `backdrop_path`, `belongs_to_collection`, `homepage`, `tagline`, and `certification` columns.

There are some missing values in `overview`, `poster_path`, and `release_date` columns.

## 3.3 Check for unnecessary columns

In [8]:
df.head()

adult                     backdrop_path  \
0    NaN                               NaN   
1    0.0  /vMFs7nw6P0bIV1jDsQpxAieAVnH.jpg   
2    0.0                               NaN   
3    0.0                               NaN   
4    0.0  /n4GJFGzsc7NinI1VeGDXIcQjtU2.jpg   

                               belongs_to_collection      budget  \
0                                                NaN         NaN   
1                                                NaN  10000000.0   
2                                                NaN         0.0   
3                                                NaN         0.0   
4  {'id': 1131062, 'name': 'Wong Kar-Wai’s Love T...    150000.0   

                                              genres homepage        id  \
0                                                NaN      NaN       NaN   
1  [{'id': 35, 'name': 'Comedy'}, {'id': 10402, '...      NaN   62127.0   
2           [{'id': 878, 'name': 'Science Fiction'}]      NaN  110977.0   
3  [{'id': 18, 'name': 'Drama'}, {'id': 28, 'name...      NaN  442869.0   
4  [{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...      NaN     843.0   

  original_language   original_title  \
0               NaN              NaN   
1                en  The Fantasticks   
2                en    For the Cause   
3                hi             Gang   
4                cn             花樣年華   

                                            overview  ...     revenue runtime  \
0                                                NaN  ...         NaN     NaN   
1  Two rural teens sing and dance their way throu...  ...         0.0    86.0   
2  Earth is in a state of constant war and two co...  ...         0.0   100.0   
3  After falling prey to underworld, four friends...  ...         0.0   152.0   
4  Two neighbors become intimate after discoverin...  ...  14204632.0    99.0   

                                    spoken_languages    status  \
0                                                NaN       NaN   
1  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
2  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
3  [{'english_name': 'Hindi', 'iso_639_1': 'hi', ...  Released   
4  [{'english_name': 'Cantonese', 'iso_639_1': 'c...  Released   

                                             tagline                 title  \
0                                                NaN                   NaN   
1      Try to remember the first time magic happened       The Fantasticks   
2       The ultimate showdown on a forbidden planet.         For the Cause   
3                                                NaN                  Gang   
4  Feel the heat, keep the feeling burning, let t...  In the Mood for Love   

   video vote_average vote_count certification  
0    NaN          NaN        NaN           NaN  
1    0.0          5.5       22.0           NaN  
2    0.0          5.1        8.0           NaN  
3    0.0          4.0        1.0           NaN  
4    0.0          8.1     2244.0            PG  

[5 rows x 25 columns]

In [9]:
#check what % of the data are missing for each column
df.isna().sum() / df.shape[0]

adult                    0.000016
backdrop_path            0.352369
belongs_to_collection    0.932565
budget                   0.000016
genres                   0.000016
homepage                 0.760134
id                       0.000016
original_language        0.000016
original_title           0.000016
overview                 0.020641
popularity               0.000016
poster_path              0.078128
production_companies     0.000016
production_countries     0.000016
release_date             0.017156
revenue                  0.000016
runtime                  0.000016
spoken_languages         0.000016
status                   0.000016
tagline                  0.617501
title                    0.000016
video                    0.000016
vote_average             0.000016
vote_count               0.000016
certification            0.758550
dtype: float64

In [10]:
#check the number of unique values in each column
df.nunique()

adult                        2
backdrop_path            40879
belongs_to_collection     2210
budget                    1660
genres                    3944
homepage                 14983
id                       63126
original_language          125
original_title           60339
overview                 61793
popularity               14308
poster_path              58192
production_companies     34910
production_countries      3430
release_date              7829
revenue                   7598
runtime                    240
spoken_languages          2672
status                       3
tagline                  23819
title                    57861
video                        2
vote_average              3518
vote_count                3059
certification               12
dtype: int64

In [11]:
#Display data under this column to determine whether to keep this column or not
#df['belongs_to_collection'].value_counts()

In [12]:
#Display data under this column to determine whether to keep this column or not
#df['homepage'].value_counts()

In [13]:
#Display data under this column to determine whether to keep this column or not
#df['tagline'].value_counts()

**Columns to be dropped due to too many missing data:**
- `backdrop_path` 
   - Reason: More than 35% of the data is missing, and this data does not look useful for hypothesis testing.
- `homepage`
   - Reason: More than 76% of the data is missing, and I will not use this data for hypothesis testing.
   
**Columns to be dropped due to too many unique values:**
- `poster_path`
   - Reason: I will not use this data for hypothesis testing.
        
**Columns with many missing data and unique values, but keeping as they may be useful:**
- `tagline`
   - Reason: More than 61.7% of the data is missing, but it may be utilized for comparing some target between two groups: (a) movies with tagline and (b) movies without tagline. Alternatively, it could be used when checking the revenue of movies that has a specific word in the tagline vs. movies that do not have the specific word in the tagline.
        
- `belongs_to_collection`
   - Reason: It is missing 93% of the data, but the feature looks useful.

- `certification`
   - Reason: It is missing more than 75% of the data, but we **need** to keep this column for our Hypothesis Testing requested by the Stakeholder.

Other columns with many unique values that are not mentioned above may also be useful, therefore, I will keep them in the dataframe for now.


In [14]:
df = df.drop(columns = ['backdrop_path', 'homepage', 'poster_path'])

In [15]:
#confirm the columns have been dropped
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 63127 entries, 0 to 63147
Data columns (total 22 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  63126 non-null  float64
 1   belongs_to_collection  4257 non-null   object 
 2   budget                 63126 non-null  float64
 3   genres                 63126 non-null  object 
 4   id                     63126 non-null  float64
 5   original_language      63126 non-null  object 
 6   original_title         63126 non-null  object 
 7   overview               61824 non-null  object 
 8   popularity             63126 non-null  float64
 9   production_companies   63126 non-null  object 
 10  production_countries   63126 non-null  object 
 11  release_date           62044 non-null  object 
 12  revenue                63126 non-null  float64
 13  runtime                63126 non-null  float64
 14  spoken_languages       63126 non-null  object 
 15  st

## 3.4 Check for inconsistent values

In [16]:
df.describe(include=object)

belongs_to_collection  \
count                                                4257   
unique                                               2210   
top     {'id': 39199, 'name': 'Detective Conan Collect...   
freq                                                   18   

                               genres original_language     original_title  \
count                           63126             63126              63126   
unique                           3944               125              60339   
top     [{'id': 18, 'name': 'Drama'}]                en  A Christmas Carol   
freq                            10827             37376                  9   

                                                 overview  \
count                                               61824   
unique                                              61793   
top     The story follows a boy named Quon and others ...   
freq                                                    5   

       production_companies  \
count                 63126   
unique                34910   
top                      []   
freq                  20911   

                                     production_countries release_date  \
count                                               63126        62044   
unique                                               3430         7829   
top     [{'iso_3166_1': 'US', 'name': 'United States o...   2009-01-01   
freq                                                16983          258   

                                         spoken_languages    status  \
count                                               63126     63126   
unique                                               2672         3   
top     [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
freq                                                23998     63112   

                      tagline  title certification  
count                   24146  63126         15242  
unique                  23819  57861            12  
top     Based on a true story   Home             R  
freq                        9     13          6252

Many of these columns have many unique values, therefore, I will focus on only checking the columns with few unique values at this point, and will clean the other columns when there is a use-case.

Columns with few unique values:
    
    - 'status' (number of unique values: 3)
    - 'certification' (number of unique values: 12)

In [17]:
df['status'].value_counts()

Released           63112
Post Production       11
In Production          3
Name: status, dtype: int64

This column does not have any inconsistent values. I would need to check with a Subject Matter Expert if I want to find out whether there is a difference in the above 3 categories.

In [18]:
df['certification'].value_counts()

R                                  6252
NR                                 3596
PG-13                              3288
PG                                 1472
G                                   457
NC-17                               167
Unrated                               5
UR                                    1
Not Rated                             1
ScreamFest Horror Film Festival       1
PG-13                                 1
10                                    1
Name: certification, dtype: int64

**Observation: This column contains many inconsistent values, therefore I will clean them.**

### 3.4.1 Cleaning inconsistent values in **`certification`** 

According to several sources (example: [www.filmratings.com](https://www.filmratings.com/)), the MPAA rating consists of 5 values + 1 ("Unrated/Not Rated") categories:
- G
- PG
- PG-13
- R
- NC-17
- Unrated / Not Rated / NR / UR
    
    

Therefore, I will take the following actions to clean up the 'certification' column:

|current value | count | replace with|
| - | - | - |
| R | 6252 | R |                              
| NR   | 3596 | Unrated |                             
| PG-13 | 3288 | PG-13 |                           
| PG    | 1472| PG |                          
| G     | 457 | G |                         
| NC-17 | 167| NC-17 |                         
| Unrated | 5 | Unrated  |                            
| UR   | 1 | Unrated |         
|Not Rated | 1 | Unrated |                            
| ScreamFest Horror Film Festival  |   1 | Drop |
| PG-13 | 1 | PG-13 |                                
| 10  | 1 | Drop |                               

In [19]:
#Replace
df['certification'].replace({'NR':'Unrated', 
                             'PG-13 ': 'PG-13', 
                             'UR' : 'Unrated', 
                             'Not Rated' : 'Unrated'
                             }, inplace=True)

In [20]:
#Filter

filter_1 = df['certification'] == 'ScreamFest Horror Film Festival'
df[filter_1]

adult belongs_to_collection  budget  \
8996    0.0                   NaN     0.0   

                                                 genres        id  \
8996  [{'id': 27, 'name': 'Horror'}, {'id': 35, 'nam...  310725.0   

     original_language     original_title  \
8996                en  Hoodoo for Voodoo   

                                               overview  popularity  \
8996  After winning an all-expenses-paid Mardis Gras...         1.4   

                                   production_companies  ... revenue runtime  \
8996  [{'id': 81428, 'logo_path': None, 'name': 'Aby...  ...     0.0    79.0   

                                       spoken_languages    status  \
8996  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                       tagline              title video vote_average  \
8996  Let the dead times roll!  Hoodoo for Voodoo   0.0          1.5   

      vote_count                    certification  
8996         2.0  ScreamFest Horror Film Festival  

[1 rows x 22 columns]

In [21]:
#Filter
filter_2 = df['certification'] == '10'
df[filter_2]

adult belongs_to_collection  budget  \
32094    0.0                   NaN     0.0   

                                                  genres        id  \
32094  [{'id': 27, 'name': 'Horror'}, {'id': 53, 'nam...  402515.0   

      original_language  original_title  \
32094                en  American Beast   

                                                overview  popularity  \
32094  After his mother's death, James Erikson discov...       3.226   

                                    production_companies  ... revenue runtime  \
32094  [{'id': 44632, 'logo_path': '/neCKH4sJCBhZ1B4Q...  ...     0.0    89.0   

                                        spoken_languages    status tagline  \
32094  [{'english_name': 'English', 'iso_639_1': 'en'...  Released     NaN   

                title video vote_average  vote_count  certification  
32094  American Beast   0.0          6.6        31.0             10  

[1 rows x 22 columns]

In [22]:
combined_filter = (filter_1) | (filter_2)

In [23]:
df[combined_filter]

adult belongs_to_collection  budget  \
8996     0.0                   NaN     0.0   
32094    0.0                   NaN     0.0   

                                                  genres        id  \
8996   [{'id': 27, 'name': 'Horror'}, {'id': 35, 'nam...  310725.0   
32094  [{'id': 27, 'name': 'Horror'}, {'id': 53, 'nam...  402515.0   

      original_language     original_title  \
8996                 en  Hoodoo for Voodoo   
32094                en     American Beast   

                                                overview  popularity  \
8996   After winning an all-expenses-paid Mardis Gras...       1.400   
32094  After his mother's death, James Erikson discov...       3.226   

                                    production_companies  ... revenue runtime  \
8996   [{'id': 81428, 'logo_path': None, 'name': 'Aby...  ...     0.0    79.0   
32094  [{'id': 44632, 'logo_path': '/neCKH4sJCBhZ1B4Q...  ...     0.0    89.0   

                                        spoken_languages    status  \
8996   [{'english_name': 'English', 'iso_639_1': 'en'...  Released   
32094  [{'english_name': 'English', 'iso_639_1': 'en'...  Released   

                        tagline              title video vote_average  \
8996   Let the dead times roll!  Hoodoo for Voodoo   0.0          1.5   
32094                       NaN     American Beast   0.0          6.6   

       vote_count                    certification  
8996          2.0  ScreamFest Horror Film Festival  
32094        31.0                               10  

[2 rows x 22 columns]

In [24]:
df = df[~combined_filter]

In [25]:
df['certification'].value_counts()

R          6252
Unrated    3603
PG-13      3289
PG         1472
G           457
NC-17       167
Name: certification, dtype: int64

## 3.5 Check for odd values

In [26]:
df.describe()

adult        budget            id    popularity       revenue  \
count  63124.000000  6.312400e+04  6.312400e+04  63124.000000  6.312400e+04   
mean       0.001838  3.038191e+06  3.814483e+05      4.893317  8.028349e+06   
std        0.042829  1.584748e+07  2.708746e+05     16.391970  6.008496e+07   
min        0.000000  0.000000e+00  1.200000e+01      0.000000  0.000000e+00   
25%        0.000000  0.000000e+00  1.273785e+05      0.677000  0.000000e+00   
50%        0.000000  0.000000e+00  3.672080e+05      1.655000  0.000000e+00   
75%        0.000000  0.000000e+00  5.752385e+05      4.689000  0.000000e+00   
max        1.000000  3.790000e+08  1.152334e+06   1947.386000  2.923706e+09   

            runtime         video  vote_average    vote_count  
count  63124.000000  63124.000000  63124.000000  63124.000000  
mean      91.840489      0.007826      4.600025    219.842073  
std       31.022150      0.088118      2.623544   1170.382903  
min        0.000000      0.000000      0.000000      0.000000  
25%       84.000000      0.000000      3.200000      1.000000  
50%       93.000000      0.000000      5.455000      6.000000  
75%      105.000000      0.000000      6.400000     35.000000  
max     1440.000000      1.000000     10.000000  33995.000000

**Observation**: There are no obvious odd values based on the above check.

# Hypothesis Testing

# Q1. Does the MPAA rating of a movie (G/PG/PG-13/R) affect how much revenue the movie generates?



## 1.  Null Hypothesis and Alternative Hypothesis

**Null Hypothesis**: The MPAA rating of a movie does not affect how much revenue the movie generates.
    
**Alternative Hypothesis**: The MPAA rating of a movie affects how much revenue the movie generates.

## 2. Selecting the Hypothesis Test

In [27]:
df['certification'].value_counts()

R          6252
Unrated    3603
PG-13      3289
PG         1472
G           457
NC-17       167
Name: certification, dtype: int64

* Target: 'revenue' - **Numerical**
* Number of groups: **6 groups**

**Test Type**: One-Way ANOVA and/or Tukey (if assumptions are met.)